source/wifi/wifi_hal.c

```diff
@@ -2995,14 +3091,206 @@ INT wifi_getAllAssociatedDeviceDetail(INT apIndex, ULONG *output_ulong, wifi_dev
 {
-	if (NULL == output_struct) {
-		return RETURN_ERR;
-	} else {
-		memset(output_struct, 0, sizeof(wifi_device_t));
-		return RETURN_OK;
-	}
+#ifdef HAL_NETLINK_IMPL
+    Netlink nl;
+    char  if_name[10];
+
+    wifi_device_info_t info;
+    info.wifi_devIndex = devIndex;
+
+    snprintf(if_name,sizeof(if_name),"wlan%d",apIndex);
+
+    nl.id = initSock80211(&nl);
+
+    if (nl.id < 0) {
+        fprintf(stderr, "Error initializing netlink \n");
+        return -1;
+    }
+
+    struct nl_msg* msg = nlmsg_alloc();
+
+    if (!msg) {
+        fprintf(stderr, "Failed to allocate netlink message.\n");
+        nlfree(&nl);
+        return -2;
+    }
+
+    genlmsg_put(msg,
+                NL_AUTO_PORT,
+                NL_AUTO_SEQ,
+                nl.id,
+                0,
+                NLM_F_DUMP,
+                NL80211_CMD_GET_STATION,
+                0);
+
+    nla_put_u32(msg, NL80211_ATTR_IFINDEX, if_nametoindex(if_name));
+    nl_send_auto(nl.socket, msg);
+    nl_cb_set(nl.cb,NL_CB_VALID,NL_CB_CUSTOM,AssoDevInfo_callback,&info);
+    nl_recvmsgs(nl.socket, nl.cb);
+    nlmsg_free(msg);
+    nl_cb_put(nl.cb);
+    nlfree(&nl);
+
+    output_struct->wifi_devAssociatedDeviceAuthentiationState = (wifi_device_t*)info.wifi_devAssociatedDeviceAuthentiationState;
+    output_struct->wifi_devRxRate = (wifi_device_t*)info.wifi_devRxRate;
+    output_struct->wifi_devTxRate = (wifi_device_t*)info.wifi_devTxRate;
+    output_struct->wifi_devSignalStrength = (wifi_device_t*)info.wifi_devSignalStrength;
+    memcpy(&output_struct->wifi_devMacAddress,&info.wifi_devMacAddress,sizeof(wifi_device_t));
+    return RETURN_OK;
+#else
+    //iw utility to retrieve station information
+    FILE *file = NULL;
+    char if_name[10] = {'\0'};
+    char pipeCmd[256] = {'\0'};
+    char line[256];
+    int count,device = 0;
+
+    snprintf(if_name,sizeof(if_name),"wlan%d",apIndex);
+
+    sprintf(pipeCmd, "iw dev %s station dump | grep %s | wc -l", if_name, if_name);
+    file = popen(pipeCmd, "r");
+    fgets(line, sizeof line, file);
+    device = atoi(line);
+
+    if(device == 0)
+        return RETURN_ERR; //No devices are connected
+
+    sprintf(pipeCmd,"iw dev %s station dump > /tmp/AssociatedDevice_Stats.txt", if_name);
+    system(pipeCmd);
+
+    system("cat /tmp/AssociatedDevice_Stats.txt | grep 'signal avg' | cut -d ' ' -f2 | cut -d ':' -f2 | cut -f 2 | tr -s '\n' > /tmp/wifi_signalstrength.txt");
+
+    system("cat /tmp/AssociatedDevice_Stats.txt | grep Station | cut -d ' ' -f 2  > /tmp/wifi_AssoMac.txt");
+
+    system("cat /tmp/AssociatedDevice_Stats.txt | grep 'tx bitrate' | cut -d ' ' -f2 | cut -d ':' -f2 |  cut -f 2 | tr -s '\n' | cut -d '.' -f1 > /tmp/wifi_txrate.txt");
+
+    system("cat /tmp/AssociatedDevice_Stats.txt | grep 'rx bitrate' | cut -d ' ' -f2 | cut -d ':' -f2 |  cut -f 2 | tr -s '\n' | cut -d '.' -f1 > /tmp/wifi_rxrate.txt");
+
+    //devIndex starts from 1, ++count
+    file = fopen("/tmp/wifi_signalstrength.txt", "r");
```

Comment:
I would like have errror handling like this 
    file = fopen("", ...);
    if(file==NULL)
    {
        perror("fopen failed");
        return -1;
    }